File: tools/typescript/src/shared/toolkit-core.ts
import {StripeMcpClient, type McpTool} from './mcp-client';
import {MultiMcpClient} from './multi-mcp-client';
import {AsyncInitializer} from './async-initializer';
import {type Configuration} from './configuration';

export type {McpTool};

/**
 * Configuration options for toolkit initialization.
 */
export interface ToolkitConfig {
  secretKey: string;
  configuration: Configuration;
}

/**
 * Base class for all Stripe Agent Toolkit implementations.
 * Subclasses override convertTools() to transform MCP tools into framework-specific formats.
 *
 * Supports additional MCP servers whose tools are merged with the core Stripe
 * tools. When a tool call is invoked, it is automatically routed to the
 * server that provides it.
 *
 * @template T - The type of tools returned by this toolkit (e.g., ChatCompletionTool[], StripeTool[])
 */
export class ToolkitCore<T = McpTool[]> {
  /**
   * The MCP client that handles connections to mcp.stripe.com and tool execution.
   */
  readonly mcpClient: StripeMcpClient;

  /**
   * Optional client for additional MCP servers (e.g. batch payment providers).
   */
  private additionalClient: MultiMcpClient | null = null;

  readonly configuration: Configuration;
  private _initializer = new AsyncInitializer();
  private _tools: T;

  constructor(config: ToolkitConfig, emptyTools: T) {
    this.mcpClient = new StripeMcpClient({
      secretKey: config.secretKey,
      context: {
        account: config.configuration.context?.account,
        customer: config.configuration.context?.customer,
      },
      mode: config.configuration.context?.mode,
    });
    this.configuration = config.configuration;
    this._tools = emptyTools;

    if (
      config.configuration.additionalMcpServers &&
      config.configuration.additionalMcpServers.length > 0
    ) {
      this.additionalClient = new MultiMcpClient();
    }
  }

  /**
   * Convert MCP tools to framework-specific format.
   * Override this method in subclasses to provide custom conversion.
   * Default implementation returns tools as-is (cast to T).
   */
  protected convertTools(mcpTools: McpTool[]): T {
    return mcpTools as unknown as T;
  }

  /**
   * Initialize the toolkit by connecting to the MCP server and fetching tools.
   * The server filters tools based on RAK permissions.
   *
   * If additional MCP servers are configured, their tools are fetched and
   * merged with the core Stripe tools.
   */
  async initialize(): Promise<void> {
    await this._initializer.initialize(async () => {
      await this.mcpClient.connect();

      const remoteTools = this.mcpClient.getTools();
      let allTools = [...remoteTools];

      // Connect additional MCP servers and merge their tools.
      // Stripe tool names are reserved — additional servers cannot shadow them.
      if (
        this.additionalClient &&
        this.configuration.additionalMcpServers
      ) {
        const stripeToolNames = remoteTools.map((t) => t.name);
        this.additionalClient.setReservedNames(stripeToolNames);

        await this.additionalClient.connect(
          this.configuration.additionalMcpServers
        );
        const additionalTools = this.additionalClient.getTools();
        allTools = [...allTools, ...additionalTools];
      }

      this._tools = this.convertTools(allTools);
    });
  }

  /**
   * Check if the toolkit has been initialized.
   */
  isInitialized(): boolean {
    return this._initializer.isInitialized;
  }

  /**
   * Get tools, throwing if not initialized.
   */
  getTools(): T {
    this.ensureInitialized();
    return this._tools;
  }

  /**
   * Get tools with a warning if not initialized.
   * @deprecated Use getTools() after calling initialize().
   */
  getToolsWithWarning(): T {
    this.warnIfNotInitialized();
    return this._tools;
  }

  /**
   * Route a tool call to the correct MCP server and return the result.
   * Core Stripe tools always take priority. Additional server tools are
   * only called for names that do not exist in the Stripe tool set.
   */
  async routeToolCall(
    name: string,
    args: Record<string, unknown>,
    options?: {customer?: string}
  ): Promise<string> {
    this.ensureInitialized();

    // Additional servers only handle tools that Stripe does not own.
    // The collision guard in MultiMcpClient already prevents shadowing,
    // but this ordering provides defense-in-depth.
    if (this.additionalClient && this.additionalClient.hasTool(name)) {
      return this.additionalClient.callTool(name, args);
    }

    // Default: route to the primary Stripe MCP server
    return this.mcpClient.callTool(name, args, options);
  }

  /**
   * Close the MCP connection and clean up resources.
   */
  async close(emptyTools: T): Promise<void> {
    if (!this._initializer.isInitialized) {
      return;
    }

    await this.mcpClient.disconnect();

    if (this.additionalClient) {
      await this.additionalClient.disconnect();
    }

    this._initializer.reset();
    this._tools = emptyTools;
  }

  /**
   * Throw an error if not initialized.
   */
  ensureInitialized(): void {
    if (!this._initializer.isInitialized) {
      throw new Error(
        'StripeAgentToolkit not initialized. ' +
          'Use `await createStripeAgentToolkit()` factory (recommended) or call `await toolkit.initialize()` first. ' +
          'See migration guide: https://github.com/stripe/agent-toolkit/blob/main/tools/typescript/MIGRATION.md'
      );
    }
  }

  /**
   * Warn if accessing tools before initialization.
   */
  warnIfNotInitialized(): void {
    if (!this._initializer.isInitialized) {
      console.warn(
        '[StripeAgentToolkit] Accessing tools before initialization. ' +
          'Use createStripeAgentToolkit() factory (recommended) or call await toolkit.initialize() first. ' +
          'Tools will be empty until initialized.'
      );
    }
  }
}

export default ToolkitCore;

File: tools/python/stripe_agent_toolkit/shared/toolkit_core.py
"""Base class for all Stripe Agent Toolkit implementations."""

from abc import ABC, abstractmethod
from typing import TypeVar, Generic, List, Optional
import warnings

from .stripe_client import StripeClient
from .mcp_client import McpTool
from .async_initializer import AsyncInitializer
from ..configuration import Configuration, is_tool_allowed_by_name

T = TypeVar("T")


class ToolkitCore(ABC, Generic[T]):
    """
    Base class for all Stripe Agent Toolkit implementations.

    Subclasses override _convert_tools() to transform MCP tools
    into framework-specific formats.

    Example:
        class MyToolkit(ToolkitCore[List[MyTool]]):
            def _empty_tools(self) -> List[MyTool]:
                return []

            def _convert_tools(self, mcp_tools: List[McpTool]) -> List[MyTool]:
                return [MyTool(t) for t in mcp_tools]

        toolkit = MyToolkit('rk_test_...', configuration)
        await toolkit.initialize()
        tools = toolkit.get_tools()
        await toolkit.close()
    """

    def __init__(
        self,
        secret_key: str,
        configuration: Optional[Configuration] = None
    ):
        self._configuration = configuration or {}
        self._stripe = StripeClient(
            secret_key,
            self._configuration.get("context")
        )
        self._initializer = AsyncInitializer()
        self._tools: T = self._empty_tools()

    @abstractmethod
    def _empty_tools(self) -> T:
        """
        Return the empty value for tools (e.g., [], {}).
        Called during initialization before tools are loaded.
        """
        pass

    @abstractmethod
    def _convert_tools(self, mcp_tools: List[McpTool]) -> T:
        """
        Convert MCP tools to framework-specific format.

        Args:
            mcp_tools: List of tools from MCP server

        Returns:
            Framework-specific tool collection
        """
        pass

    async def initialize(self) -> None:
        """
        Initialize the toolkit by connecting to MCP server and fetching tools.

        This must be called before using get_tools() or running tool calls.
        """
        await self._initializer.initialize(self._do_initialize)

    async def _do_initialize(self) -> None:
        """Internal initialization logic."""
        await self._stripe.initialize()

        remote_tools = self._stripe.get_remote_tools()
        filtered_tools = [
            t for t in remote_tools
            if is_tool_allowed_by_name(t.get("name", ""), self._configuration)
        ]

        self._tools = self._convert_tools(filtered_tools)

    @property
    def is_initialized(self) -> bool:
        """Check if toolkit is initialized."""
        return self._initializer.is_initialized

    def get_tools(self) -> T:
        """
        Get tools, throwing if not initialized.

        Raises:
            RuntimeError: If initialize() has not been called.
        """
        self._ensure_initialized()
        return self._tools

    def _get_tools_with_warning(self) -> T:
        """
        Get tools with a warning if not initialized.
        Used for deprecated property access.
        """
        self._warn_if_not_initialized()
        return self._tools

    async def close(self) -> None:
        """
        Close the MCP connection and clean up resources.
        Safe to call multiple times.
        """
        if not self._initializer.is_initialized:
            return

        await self._stripe.close()
        self._initializer.reset()
        self._tools = self._empty_tools()

    def _ensure_initialized(self) -> None:
        """Throw an error if not initialized."""
        if not self._initializer.is_initialized:
            raise RuntimeError(
                "StripeAgentToolkit not initialized. "
                "Call await toolkit.initialize() first."
            )

    def _warn_if_not_initialized(self) -> None:
        """Warn if accessing tools before initialization."""
        if not self._initializer.is_initialized:
            warnings.warn(
                "[StripeAgentToolkit] Accessing tools before initialization. "
                "Call await toolkit.initialize() first, or use "
                "create_stripe_agent_toolkit() factory. "
                "Tools will be empty until initialized."
            )

    @property
    def stripe(self) -> StripeClient:
        """
        Access to the underlying StripeClient.
        Useful for billing operations like create_meter_event().
        """
        return self._stripe
